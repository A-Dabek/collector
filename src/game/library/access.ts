import { rarities, Rarity } from '../../ui/rarity';
import { LIBRARY } from './library';

export interface Card {
  id: number;
  name: string;
  rarity: Rarity;
  cost: number;
  description: string;
  enabled: boolean;
}

export function findCardInLibrary(name: string): Card {
  const libraryCard = LIBRARY[name];
  return {
    id: Math.random(),
    name,
    cost: rarities.findIndex((item) => item === libraryCard.rarity) + 1,
    enabled: true,
    ...libraryCard,
  };
}

export function getRandomCard(): Card {
  const rarity = rarities[getRarity()];
  const randomKey = findRandomCardOfRarity(rarity);
  return findCardInLibrary(randomKey);
}

function findRandomCardOfRarity(rarity: Rarity): string {
  const cardNames = Object.keys(LIBRARY).filter(
    (key) => LIBRARY[key].rarity === rarity,
  );
  const randomIndex = Math.floor(Math.random() * cardNames.length);
  return cardNames[randomIndex];
}

function getRarity(): number {
  const decreasingFactor = 10;

  const calculateBaseProbability = (factor: number): number => {
    const sumInversePowers = Array(6)
      .fill(0)
      .reduce((sum, _, i) => sum + 1 / Math.pow(factor, i), 0);
    return 1 / sumInversePowers;
  };

  const normalizeProbabilities = (factor: number): number[] => {
    const baseProbability = calculateBaseProbability(factor);
    return Array(6)
      .fill(0)
      .map((_, i) => baseProbability / Math.pow(factor, i));
  };

  const probabilities = normalizeProbabilities(decreasingFactor);
  const randomValue = Math.random();
  let cumulativeProbability = 0;

  for (let i = 0; i < probabilities.length; i++) {
    cumulativeProbability += probabilities[i];
    if (randomValue <= cumulativeProbability) {
      return i;
    }
  }

  return 0; // Default case, though it shouldn't happen
}
