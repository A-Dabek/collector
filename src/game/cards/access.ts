import { GameCard } from './card';
import { CardPickupCard } from './common/card-pickup.card';
import { PaperBagOpenCard } from './common/paper-bag-open.card';
import { PillCard } from './common/pill.card';
import { SpikesInitCard } from './common/spikes-init.card';
import { TokenCard } from './common/token.card';
import { PiggyBankCard } from './piggy-bank.card';
import { PlainPadlockCard } from './plain-padlock.card';
import { Card2DiamondsCard } from './uncommon/card-2-diamonds.card';
import { HammerBreakCard } from './uncommon/hammer-break.card';
import { MedicinesCard } from './uncommon/medicines.card';
import { PaperBagCrumpledCard } from './uncommon/paper-bag-crumpled.card';
import { ShoulderBagCard } from './uncommon/shoulder-bag.card';
import { SpikesHalfCard } from './uncommon/spikes-half.card';
import { TwoCoinsCard } from './uncommon/two-coins.card';
import { RaiseZombieCard } from './raise-zombie.card';

/*
(feasible)
- increase luck
- die - get a card of same rarity or not
- note with a sword - complete a task for points
- add 5 spaces with cards of every rarity
- play 5 random cards for free
- destroy a card and get a higher rarity card
- draw rare-less cards until space is full
- heal flat for every space available
- missing HP is max HP, full heal
- add missing health as max HP
- re-roll all cards
- shrink spaces with each distinct rarity

(needs play history)
- exhume a card
- parrot - same effect as previous

(needs targeting)
- ouroboros - play a card infinitely as long as it's possible (needs targetting)

(needs options)
- shop - buy a card for points
- goldfish/djinn - wishes

(unknown)
- fisherman - random fish?
- incognito - get a card without info
- time limit - lose hp? every second
- tick - drains life?
- chessboard - add chess pieces
*/

export type CardName =
  // common
  | 'card-pickup'
  | 'pill'
  | 'paper-bag-open'
  | 'spikes-init'
  | 'token'
  | 'plain-padlock'
  | 'piggy-bank' //FIXME not common

  // uncommon
  | 'paper-bag-crumpled'
  | 'hammer-break'
  | 'card-2-diamonds'
  | 'medicines'
  | 'shoulder-bag'
  | 'spikes-half'
  | 'two-coins'
  | 'raise-zombie';

export const PLAYABLE_LIBRARY: Record<CardName, new () => GameCard> = {
  //common
  'card-pickup': CardPickupCard,
  pill: PillCard,
  'paper-bag-open': PaperBagOpenCard,
  'spikes-init': SpikesInitCard,
  token: TokenCard,
  'piggy-bank': PiggyBankCard,
  'plain-padlock': PlainPadlockCard,

  //uncommon
  'paper-bag-crumpled': PaperBagCrumpledCard,
  'card-2-diamonds': Card2DiamondsCard,
  medicines: MedicinesCard,
  'shoulder-bag': ShoulderBagCard,
  'spikes-half': SpikesHalfCard,
  'two-coins': TwoCoinsCard,
  'hammer-break': HammerBreakCard,
  'raise-zombie': RaiseZombieCard,
};
