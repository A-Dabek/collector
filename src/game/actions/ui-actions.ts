import { CardState } from '../cards/card';
import { EffectState } from '../effects/effect';

export interface GameUiState {
  points: number;
  maxPoints: number;
  health: number;
  maxHealth: number;
  space: number;
  cards: CardState[];
  effects: EffectState[];
}
