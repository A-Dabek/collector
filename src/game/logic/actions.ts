import { ItemId } from '../../services/collection-persistence.service';
import { CollectionRollService } from '../../services/collection-roll.service';
import {
  PersistenceAction,
  PersistenceActionFactory,
} from '../../services/persistence-action.factory';
import { itemAddUiAction, pointsChangeUiAction, UiAction } from '../ui-actions';

export interface ResponseActions {
  persistenceActions: PersistenceAction[];
  uiActions: UiAction[];
}

export function combineActions(actions: ResponseActions[]): ResponseActions {
  const combinedPersistenceActions = actions.flatMap(
    (action) => action.persistenceActions,
  );
  const combinedUiActions = actions.flatMap((action) => action.uiActions);

  return {
    persistenceActions: combinedPersistenceActions,
    uiActions: combinedUiActions,
  };
}

export function pointsChangeAction(payload: number): ResponseActions {
  return {
    uiActions: [pointsChangeUiAction(payload)],
    persistenceActions: [PersistenceActionFactory.setPoints(payload)],
  };
}

export function itemAddAction(payload: ItemId): ResponseActions {
  return {
    uiActions: [itemAddUiAction(payload)],
    persistenceActions: [PersistenceActionFactory.setItem(payload)],
  };
}

export function cardCostAction(rarity: number): ResponseActions {
  return pointsChangeAction(6 - rarity);
}

export function cardDrawAction(count: number): ResponseActions {
  const roll = new CollectionRollService();
  return itemAddAction(roll.getRandomItem());
}

export function healAction(payload: number): ResponseActions {
  return {
    uiActions: [],
    persistenceActions: [],
  };
}

export function expandAction(count: number): ResponseActions {
  return {
    uiActions: [],
    persistenceActions: [],
  };
}

export function shrinkRandomAction(count: number): ResponseActions {
  return {
    uiActions: [],
    persistenceActions: [],
  };
}

export function shrinkAction(index: number): ResponseActions {
  return {
    uiActions: [],
    persistenceActions: [],
  };
}

export function destroyAction(): ResponseActions {
  return {
    uiActions: [],
    persistenceActions: [],
  };
}

export function destroyRandomAction(count: number): ResponseActions {
  return {
    uiActions: [],
    persistenceActions: [],
  };
}

export function earnPointsAction(points: number): ResponseActions {
  return {
    uiActions: [],
    persistenceActions: [],
  };
}

export function changeMaxHpAction(change: number): ResponseActions {
  return {
    uiActions: [],
    persistenceActions: [],
  };
}

export function healMissingHpAction(percent: number): ResponseActions {
  return {
    uiActions: [],
    persistenceActions: [],
  };
}

export function increaseLuckAction(luck: number): ResponseActions {
  return {
    uiActions: [],
    persistenceActions: [],
  };
}

export function healMaxHpAction(percent: number): ResponseActions {
  return {
    uiActions: [],
    persistenceActions: [],
  };
}
