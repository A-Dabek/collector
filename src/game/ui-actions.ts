import { ItemId } from '../services/collection-persistence.service';
import { GameState } from './logic/state';

export interface GameUiState extends GameState {}

export interface UiAction {
  update: (state: GameState) => GameState;
  type?: string;
}

export function pointsChangeUiAction(payload: number): UiAction {
  return {
    type: 'pointsChange',
    update: (state) => ({
      ...state,
      points: state.points + payload,
    }),
  };
}

export function itemAddUiAction(payload: ItemId): UiAction {
  return {
    type: 'itemAdd',
    update: (state) => ({
      ...state,
      items: [...state.items, payload],
    }),
  };
}
