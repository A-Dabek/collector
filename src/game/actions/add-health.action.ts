import { GameState } from '../logic/engine';
import { GameAction } from './game-actions';
import { SetHealthAction } from './set-health.action';

export class AddHealthAction implements GameAction {
  constructor(private readonly increment: number) {}

  get description(): string {
    return `Heal (${this.increment})`;
  }

  next(state: GameState) {
    return new SetHealthAction(state.health + this.increment).next(state);
  }
}
