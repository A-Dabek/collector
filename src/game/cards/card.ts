import { Rarity } from '../../ui/rarity';
import { CardPlayAction } from '../actions/card-play.action';
import { CostHealthAction } from '../actions/cost-health.action';
import {
  GameAction,
  GameActionCreator,
  GameReactionCreator,
} from '../actions/game-actions';
import { GameState } from '../logic/engine';
import { CardName } from './access';

export interface Describable {
  name: string;
  description: string;
}

export interface CardState extends Describable {
  id: number;
  name: CardName;
  rarity: Rarity;
  enabled: boolean;
  targetSource: boolean;
  targetCandidate: boolean;
  targetDest: boolean;
}

export interface GameCard {
  id: number;
  name: CardName;
  rarity: Rarity;

  get description(): string;

  play(state: GameState): GameAction[];

  target?(state: GameState): GameAction[];

  onAction?(state: GameState, action: GameAction): GameAction[];

  onEffect?(state: GameState, actions: GameAction[]): GameAction[];

  enabled(state: GameState): boolean;

  serialize(state: GameState): CardState;
}

export abstract class BasePlayableCard implements GameCard {
  protected actions: (GameAction | GameActionCreator)[] = [];
  protected reactions: GameReactionCreator[] = [];

  abstract readonly name: CardName;
  abstract readonly rarity: Rarity;
  readonly id = Math.random();

  get description(): string {
    return [...this.costActions(), ...this.actions, ...this.reactions]
      .map((action) => action.description)
      .join(', ');
  }

  protected costActions(): GameAction[] {
    return [new CostHealthAction(this.rarity), new CardPlayAction(this)];
  }

  play(state: GameState): GameAction[] {
    return [...this.costActions(), ...this.actions].map((action) =>
      'create' in action ? action.create(state) : action,
    );
  }

  onAction(state: GameState, action: GameAction): GameAction[] {
    if (this.reactions.length === 0) return [];
    return this.reactions.flatMap((reaction) => {
      if (reaction.check(action, this)) {
        return reaction.create(state);
      }
      return [];
    });
  }

  enabled(state: GameState): boolean {
    return true;
  }

  serialize(state: GameState): CardState {
    return {
      name: this.name,
      id: this.id,
      rarity: this.rarity,
      description: this.description,
      enabled: this.enabled(state),
      targetSource: false,
      targetCandidate: false,
      targetDest: false,
    };
  }
}
