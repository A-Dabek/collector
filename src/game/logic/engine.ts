import { signal } from '@angular/core';
import { GAME_ACTIONS, ResponseActions } from '../actions/game-actions';
import { UI_ACTIONS } from '../actions/ui-actions';
import { PLAYABLE_LIBRARY } from '../cards/access';
import { NewGameCard } from '../cards/new-game.card';
import { Card } from '../library/access';
import { FinishGameCard } from '../cards/finish-game.card';
import { PlayableCard } from '../cards/card';
import { rarities, Rarity } from '../../ui/rarity';

export interface GameState {
  points: number;
  maxPoints: number;
  health: number;
  maxHealth: number;
  space: number;
  cards: Card[];
}

export class GameEngine {
  private readonly initialState: GameState = {
    points: 0,
    maxPoints: 100,
    health: 0,
    maxHealth: 100,
    space: 10,
    cards: [],
  };

  readonly state = signal<GameState>(this.initialState);

  startNewGame(): ResponseActions {
    const newGame = new NewGameCard();
    return this.playCard(newGame);
  }

  play(card: Card): ResponseActions {
    const playableCard = PLAYABLE_LIBRARY[card.name];
    return this.playCard(playableCard, card);
  }

  finishCurrentGame(): ResponseActions {
    const finishGame = new FinishGameCard();
    return this.playCard(finishGame);
  }

  private playCard(playableCard: PlayableCard, card?: Card) {
    const rarityToNumber = (rarity: Rarity): number => {
      const index = rarities.findIndex((r) => r === rarity);
      return index + 1; // Adding 1 to transform 0-based index to 1-based number
    };

    let nextState = this.state();

    let response = GAME_ACTIONS.setHealth(
      nextState.health - (card ? rarityToNumber(card.rarity) : 0),
    )(nextState);

    nextState = response.nextState;

    if (card) {
      const response2 = GAME_ACTIONS.cardWaste(card)(nextState);
      response = {
        ...response2,
        uiActions: [...response.uiActions, ...response2.uiActions],
        persistenceActions: [
          ...response.persistenceActions,
          ...response2.persistenceActions,
        ],
      };
    }

    nextState = response.nextState;

    const afterMath = playableCard.play(nextState);
    nextState = afterMath.nextState;
    nextState = this.updateCardsStatuses(nextState);
    this.state.set(nextState);
    return {
      ...afterMath,
      uiActions: [
        ...response.uiActions,
        ...afterMath.uiActions,
        UI_ACTIONS.setState(nextState),
      ],
    };
  }

  private updateCardsStatuses(state: GameState) {
    const newCards = state.cards.map((card) => {
      const playableCard = PLAYABLE_LIBRARY[card.name];
      return {
        ...card,
        enabled: playableCard.enabled(state),
      };
    });
    return { ...state, cards: newCards };
  }
}
