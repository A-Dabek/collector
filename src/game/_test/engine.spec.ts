import { Rarity } from '../../ui/rarity';
import { CardAddAction } from '../actions/basic/card-add.action';
import { SetHealthAction } from '../actions/basic/set-health.action';
import { SetSpaceAction } from '../actions/basic/set-space.action';
import { GameUiState } from '../actions/ui-actions';
import { CardName } from '../cards/access';
import { BasePlayableCard, CardState, GameCard } from '../cards/card';
import { GameEngine, GameState } from '../logic/engine';

export class TestGameCard extends BasePlayableCard {
  override name: CardName = 'card-2-diamonds';
  override rarity: Rarity = 'common';

  private get initialState(): GameState {
    return { ...GameEngine.initialState };
  }

  constructor(
    private readonly options: {
      cards?: CardName[];
    },
  ) {
    super();
  }

  override costActions() {
    return [];
  }

  override play(state: GameState) {
    return [
      new SetHealthAction(this.initialState.maxHealth),
      new SetSpaceAction(this.initialState.space),
      new CardAddAction(this.options.cards || []),
    ];
  }
}

describe('Card', () => {
  async function setup(options: { startCard: GameCard }) {
    const engine = new GameEngine();
    const snapshots = [] as GameUiState[];
    await engine.playCard(options.startCard);
    engine.snapshots$.subscribe((s) => snapshots.push(s));
    const findCardId: (name: CardName) => number | undefined = (name) =>
      snapshots.at(-1)?.cards.find((card: CardState) => card.name === name)?.id;
    const getSanitizedLastSnapshot = () => {
      const snapshot = snapshots.at(-1)!;
      return {
        ...snapshot,
        cards: snapshot.cards.map((card) => ({
          ...card,
          id: undefined,
        })),
      };
    };
    return { engine, snapshots, findCardId, getSanitizedLastSnapshot };
  }

  describe('Token', () => {
    it('use', async () => {
      const { engine, snapshots, findCardId, getSanitizedLastSnapshot } =
        await setup({
          startCard: new TestGameCard({
            cards: ['token'],
          }),
        });
      const cardId = findCardId('token')!;
      await engine.playCardId(cardId);
      expect(snapshots.length).toBe(4);
      expect(getSanitizedLastSnapshot()).toMatchSnapshot();
    });
  });
});
