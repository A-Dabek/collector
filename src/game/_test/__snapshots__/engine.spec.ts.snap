// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Token use 1`] = `
{
  "autoPlayQueue": [],
  "cardHistory": [
    "token",
  ],
  "cards": [],
  "effects": [],
  "health": 9,
  "maxHealth": 10,
  "maxPoints": 10,
  "mode": "play",
  "points": 1,
  "space": 5,
}
`;
